modules/mini-games/Race.lua

```diff
@@ -90,71 +73,109 @@ local function resetall()
         end
-    else
-        task.set_timeout_in_ticks(60, token_for_start)
     end
-end
-token_for_start = Token.register(race_count_down)
+end)
 
+--- Pick a random player who is not in start_players
 local function pick_player()
     local player = game.connected_players[math.random(#game.connected_players)]
     if not start_players[player.name] then
         return player
     else
-        return  pick_player()
+        return pick_player()
     end
 end
 
+--- Get all the script areas, and get the gates in those areas
+local function setup_areas()
+    -- Get the areas that will trigger gates to open
+    if not areas[1] then
+        areas[1] = surface[1].get_script_areas("gate_1_box")[1].area
+        areas[2] = surface[1].get_script_areas("gate_2_box")[1].area
+        areas[3] = surface[1].get_script_areas("gate_3_box")[1].area
+        areas[4] = surface[1].get_script_areas("gate_4_box")[1].area
+    end
 
-local start = function(args)
+    -- Get the areas where gates are located
+    if not gate_boxes[1] then
+        gate_boxes[1] = surface[1].get_script_areas("gate_1")[1].area
+        gate_boxes[2] = surface[1].get_script_areas("gate_2")[1].area
+        gate_boxes[3] = surface[1].get_script_areas("gate_3")[1].area
+        gate_boxes[4] = surface[1].get_script_areas("gate_4")[1].area
+    end
+
+    -- Cant correct spelling of finsish_line because it is the name of a script area
+    variables["finish"] = surface[1].get_script_areas("finsish_line")[1].area
+
+    -- Get all gates in the gate areas
+    if not gates[1] then
+        gates[1] = surface[1].find_entities_filtered{ area = gate_boxes[1], name = "gate" }
+        gates[2] = surface[1].find_entities_filtered{ area = gate_boxes[2], name = "gate" }
+        gates[3] = surface[1].find_entities_filtered{ area = gate_boxes[3], name = "gate" }
+        gates[4] = surface[1].find_entities_filtered{ area = gate_boxes[4], name = "gate" }
+    end
+end
+
+--- Function called by mini game module to start a race
+local function start(args)
     surface[1] = game.surfaces["Race game"]
-    variables["done_left"] = 0
+    variables["done_left"]  = 0
     variables["count_down"] = 4
     variables["done_right"] = 0
-    variables["left"] = true
-    variables["error_game"] = nil
-    variables["fuel"] = args[1]
-    variables["laps"] = tonumber(args[2])
-    variables["player"] = tonumber(args[3])
-    variables["place"] = 1
-    variables["new_joins"] = 0
-    scores["finshed_times"] = {}
+    variables["left"]       = true
+    variables["fuel"]       = args[1]
+    variables["laps"]       = tonumber(args[2])
+    variables["player"]     = tonumber(args[3])
+    variables["place"]      = 1
+    variables["new_joins"]  = 0
+    scores["finish_times"]  = {}
+
+    -- Error if no lap count was given
     if not variables["laps"] then
-        variables["error_game"] = "No laps set"
+        return Mini_games.error_in_game("No lap count given")
     end
```

Comment:
this will return and because of that not end the function which will give an error as the error_in_game calls the stop function

Reply:
which will try do think which cant be done (the stop func)

Reply:
okey i gues your right